controllers/helmchart_controller.go

```diff
@@ -389,18 +392,50 @@ func (r *HelmChartReconciler) reconcileFromHelmRepository(ctx context.Context,
 			return sourcev1.HelmChartNotReady(chart, sourcev1.StorageOperationFailedReason, err.Error()), err
 		}
 
-		// Find override file and retrieve contents
-		var valuesData []byte
-		cfn := filepath.Clean(chart.Spec.ValuesFile)
-		for _, f := range helmChart.Files {
-			if f.Name == cfn {
-				valuesData = f.Data
-				break
+		// Prepend a list of valuesFiles with the deprecated ValuesFile key.
+		if chart.Spec.ValuesFile != "" {
+			valuesFiles = append(valuesFiles, chart.Spec.ValuesFile)
+		}
+		valuesFiles = append(valuesFiles, chart.Spec.ValuesFiles...)
+
+		for _, v := range valuesFiles {
+			srcPath, err := securejoin.SecureJoin(tmpDir, v)
+			if err != nil {
+				return sourcev1.HelmChartNotReady(chart, sourcev1.StorageOperationFailedReason, err.Error()), err
+			}
+			if f, err := os.Stat(srcPath); os.IsNotExist(err) || !f.Mode().IsRegular() {
+				err = fmt.Errorf("invalid values file path: %s", v)
+				return sourcev1.HelmChartNotReady(chart, sourcev1.StorageOperationFailedReason, err.Error()), err
+			}
+
+			valuesData, err := ioutil.ReadFile(srcPath)
+			if err != nil {
+				err = fmt.Errorf("failed to read from values file '%s': %w", v, err)
+				return sourcev1.HelmChartNotReady(chart, sourcev1.StorageOperationFailedReason, err.Error()), err
+			}
+
+			yamlMap := make(map[string]interface{})
+			err = yaml.Unmarshal(valuesData, &yamlMap)
+			if err != nil {
+				err = fmt.Errorf("unmarshaling values from %s failed: %w", v, err)
+				return sourcev1.HelmChartNotReady(chart, sourcev1.StorageOperationFailedReason, err.Error()), err
+			}
+
+			if valuesMap == nil {
+				valuesMap = yamlMap
+			} else {
+				valuesMap = transform.MergeMaps(valuesMap, yamlMap)
 			}
 		}
 
+		yamlBytes, err := yaml.Marshal(valuesMap)
+		if err != nil {
+			err = fmt.Errorf("marshaling values failed: %w", err)
+			return sourcev1.HelmChartNotReady(chart, sourcev1.ChartPackageFailedReason, err.Error()), err
+		}
```

Comment:
This is duplicated code between the `HelmRepository` and `GitRepository` sources. Should probably be broken out into its own func?

Reply:
Most of the code here is a candidate for refactoring into more structured bits and pieces, see also: https://github.com/fluxcd/source-controller/pull/294#issuecomment-780950996

Reply:
This has diverged a bunch between `HelmRepository` and `GitRepository` sources. Now the only pieces of code shared are the creation of the yaml map and the call to `transform.MergeMaps`. I don't know if there's any real value in breaking that into its own function or if a little duplication is okay.